packages/dd-trace/test/llmobs/tagger.spec.js

```diff
@@ -0,0 +1,428 @@
+      tagger.setLLMObsSpanTags(span, 'llm')
+
+      expect(span.context()._tags).to.deep.equal({
+        'span.type': 'llm',
+        '_ml_obs.meta.span.kind': 'llm',
+        '_ml_obs.meta.ml_app': 'my-default-ml-app',
+        '_ml_obs.llmobs_parent_id': 'undefined'
+      })
+    })
+
+    it('uses the parent span if provided to populate fields', () => {
+      const parentSpan = {
+        context () {
+          return {
+            _tags: {
+              '_ml_obs.meta.ml_app': 'my-ml-app',
+              '_ml_obs.session_id': 'my-session'
+            },
+            toSpanId () { return '5678' }
+          }
+        }
+      }
+      tagger.setLLMObsSpanTags(span, 'llm', { parentLLMObsSpan: parentSpan })
+
+      expect(span.context()._tags).to.deep.equal({
+        'span.type': 'llm',
+        '_ml_obs.meta.span.kind': 'llm',
+        '_ml_obs.meta.ml_app': 'my-ml-app',
+        '_ml_obs.session_id': 'my-session',
+        '_ml_obs.llmobs_parent_id': '5678'
+      })
+    })
+
+    it('uses the propagated trace id if provided', () => {
+      tagger.setLLMObsSpanTags(span, 'llm')
+
+      expect(span.context()._tags).to.deep.equal({
+        'span.type': 'llm',
+        '_ml_obs.meta.span.kind': 'llm',
+        '_ml_obs.meta.ml_app': 'my-default-ml-app',
+        '_ml_obs.llmobs_parent_id': 'undefined'
+      })
+    })
+
+    it('uses the propagated parent id if provided', () => {
+      spanContext._trace.tags['_dd.p.llmobs_parent_id'] = '-567'
+
+      tagger.setLLMObsSpanTags(span, 'llm')
+
+      expect(span.context()._tags).to.deep.equal({
+        'span.type': 'llm',
+        '_ml_obs.meta.span.kind': 'llm',
+        '_ml_obs.meta.ml_app': 'my-default-ml-app',
+        '_ml_obs.llmobs_parent_id': '-567'
+      })
+    })
+
+    it('does not set span type if the LLMObs span kind is falsy', () => {
+      tagger.setLLMObsSpanTags(span, false)
+
+      expect(span.context()._tags['span.type']).to.be.undefined
+    })
+  })
+
+  describe('tagMetadata', () => {
+    it('tags a span with metadata', () => {
+      tagger.tagMetadata(span, { a: 'foo', b: 'bar' })
+      expect(span.context()._tags).to.deep.equal({
+        '_ml_obs.meta.metadata': '{"a":"foo","b":"bar"}'
+      })
+    })
+
+    it('logs when metadata is not JSON serializable', () => {
+      const metadata = unserializbleObject()
+      tagger.tagMetadata(span, metadata)
+      expect(logger.warn).to.have.been.calledOnce
+    })
+  })
+
+  describe('tagMetrics', () => {
+    it('tags a span with metrics', () => {
+      tagger.tagMetadata(span, { a: 1, b: 2 })
+      expect(span.context()._tags).to.deep.equal({
+        '_ml_obs.meta.metadata': '{"a":1,"b":2}'
+      })
+    })
+
+    it('logs when metrics is not JSON serializable', () => {
+      const metadata = unserializbleObject()
+      tagger.tagMetadata(span, metadata)
+      expect(logger.warn).to.have.been.calledOnce
+    })
+  })
+
+  describe('tagLLMIO', () => {
```

Comment:
These tests mostly only test that private methods are called when a public method is called. This doesn't provide a lot of confidence that the feature actually work. Could the output be tested instead? This could be the tags on the span itself, or even better, what will actually be sent (proving that the tags also properly cause the span to be split in 2 representations for tracing and llmobs).

Reply:
ah yeah, that's a good idea. i think i'll opt for checking the tags directly. i plan on adding a more integration-style/E2E testing once I have the SDK PR ready (it'll be a big one 😓)